spec/truefi2/FixedTermLoan.spec

```diff
@@ -0,0 +1,20 @@
+using ERC20 as token
+
+methods {
+    debt() returns uint256 envfree
+    unpaidDebt() returns uint256 envfree
+
+    token.balanceOf(address) returns uint256 envfree
+    
+    transfer(address, uint256) returns (bool) => DISPATCHER(true)
+    transferFrom(address, address, uint256) returns (bool) => DISPATCHER(true)
+    approve(address, uint256) returns (bool) => DISPATCHER(true)
+    unlock(address) => DISPATCHER(true)
+}
+
+invariant debtNotGreaterThanSumOfBalanceAndUnpaidDebt()
+    debt() <= token.balanceOf(currentContract) + unpaidDebt()
+    filtered { f -> 
```

Comment:
Can you explain what `invariant` and `filtered` keywords mean?

Reply:
Sure, in general, an invariant is a property that is always satisfied for a contract. Important note here is that it needs to be inductive, think of mathematical induction:
1. It needs to be satisfied initially.
2. It must be satisfied after an action assuming it's satisfied beforehand:
```
rule invariant_is_satisfied() {
  require P();
  
  method f;
  calldataarg args;
  env e;
  f(e, args);

  assert P();
}
```

Let's have an example: there is a contract with 4 expressible states - `A`, `B`, `C`, `D`. The initial state is `A` and transitions are:
`A -> B`
`B -> A`
`C -> D`
`D -> C`
This means that property `not D` is always satisfied. However, it's not a proper invariant, because state `C` satisfies the property and yet it transits to `D` which violates it. The proper invariant in this case would be `not C and not D`.


Now, the `filtered` keyword filters which functions to actually take into account when proving those transitions. The same keyword can be used in rules and it's because of how `method` variables are treated: they generate copies of the original rule with the `method` replaced with an actual function of the contract. Therefore, a `require f.selector = redeem().selector` would create a vacuous check, meaning that it would reduce to `require false`.